src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1719 @@
+  }
+  return std::move(original);
+}
+
+// Concatenate two arrays
+template <class T>
+Array<T> Concat(Array<T> a, const Array<T>& b) {
+  for (const auto& x : b) {
+    a.push_back(x);
+  }
+  return std::move(a);
+}
+
+// Combine all expressions from the container using &&.
+template <class container>
+Expr All(const container& c) {
+  Expr res;
+  for (const auto& e : c) {
+    if (res.get()) {
+      res = res && e;
+    } else {
+      res = e;
+    }
+  }
+  if (res.get()) {
+    return res;
+  } else {
+    return const_true();
+  }
+}
+
+// Create a select statement of the form cond ? on_true : 0
+Expr SelectElseZero(const Expr& cond, const Expr& on_true) {
+  return Select::make(cond, on_true, make_zero(on_true.type()));
+}
+
+// Simplify the expression as thoroughly as possible by using all available simplifiers.
+Expr SuperSimplify(Expr e, const Map<Var, Range>& vranges = Map<Var, Range>()) {
+  // For some reason no simplifier can detect that there is only one value of the variable
+  std::unordered_map<const Variable*, Expr> vmap;
+  for (const auto& var_range : vranges) {
+    if (is_const_int(var_range.second->extent, 1)) {
+      vmap[var_range.first.get()] = var_range.second->min;
+    }
+  }
+  if (!vmap.empty()) {
+    e = Substitute(e, vmap);
+  }
+
+  return CanonicalSimplify(Simplify(CanonicalSimplify(e, vranges), vranges), vranges);
+}
+
+// Provability check that uses SuperSimplify
+bool CanProve(Expr e, const Map<Var, Range>& vranges = Map<Var, Range>()) {
+  return is_one(SuperSimplify(e, vranges));
+}
+
+class ExprFreeVarsVisitor : public IRVisitor {
+ public:
+  std::vector<Var> free_array;
+  std::unordered_set<const Variable*> bound;
+  std::unordered_set<const Variable*> free;
+
+  virtual void Visit(const NodeRef& node) {
+    if (const Variable* v = node.as<Variable>()) {
+      if (!bound.count(v) && !free.count(v)) {
+        free.insert(v);
+        free_array.push_back(Var(node.node_));
+      }
+    } else {
+      IRVisitor::Visit(node);
+    }
+  }
+
+  void Visit_(const Variable* op) {
+    CHECK(false) << "This case shouldn't happen";
+  }
+
+  void Visit_(const LetStmt* op) {
+    bound.insert(op->var.get());
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const For* op) {
+    bound.insert(op->loop_var.get());
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Let* op) {
+    bound.insert(op->var.get());
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Reduce* op) {
+    for (const auto& iv : op->axis) {
+      bound.insert(iv->var.get());
+    }
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Store* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Allocate* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Free* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Load* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+};
```

Comment:
If we want a compile time error/warning then the simplest solution is to create an alternative of `IRVisitor` with every `Visit_` function declared as purely virtual:
```C++
class IRVisitorNoDefault : public IRVisitor {
 public:
  virtual void Visit_(const Variable* op) = 0;
  virtual void Visit_(const LetStmt* op) = 0;
  //...
};
```
(Using a switch and configuring the compiler to raise and error on an incomplete switch could be an alternative, but it may be a bit compiler-dependent).

However, I'm not sure if it's worth doing since new operations are rarely added into IR.